README.md
Date enhancer
===============
> The JavaScript Date enhancement

[![Build Status](https://travis-ci.org/Tiross/date-enhancer.svg?branch=master)](https://travis-ci.org/Tiross/date-enhancer)
[![Coverage Status](https://coveralls.io/repos/Tiross/date-enhancer/badge.svg?branch=master)](https://coveralls.io/r/Tiross/date-enhancer?branch=master)
[![Code Climate](https://codeclimate.com/github/Tiross/date-enhancer/badges/gpa.svg)](https://codeclimate.com/github/Tiross/date-enhancer)


How to use it ?
------------------

You'll find two file in the build directory.

Copy them to your JavaScript folder and add `<script src="js-folder/date-enhancer.min.js"></script>` in your HTML file.


What's new ?
---------------

### isValid


Check if a created instance of `Date` is a valid date.

```javascript
var validDate = new Date(2015, 2, 1);

if (validDate.isValid()) {
  // It's true !
}

var invalidDate = new Date('lol');

if (invalidDate.isValid()) {
  alert('this will never be shown');
}
```

### getWeekNumber


Return the year and week number in the format `YYYYWW`.

```javascript
var firstDayOf2016 = new Date(2016, 0);

alert(firstDayOf2016.getWeekNumber()); // will show 201553
```

### format


`format` is a port of `date` function of PHP.

You can use the same format pattern as defined on [PHP.net](http://php.net/manual/en/function.date.php).

```javascript
var date = new Date(2015, 4, 4);

alert(date.format('F \\t\\h\\e jS \\b\\e \\y\\o\\u')); // Star Wars day :)
```

The following characters are recognized in the format parameter string.


#### Day

character | Description | Example
:---------------:|-------------|--------
d | Day of the month, 2 digits with leading zeros | 01 to 31
D | A textual representation of a day, three letters | Mon through Sun
j | Day of the month without leading zeros | 1 to 31
l<br>lowercase "L" | A full textual representation of the day of the week | Sunday through Saturday
N | ISO-8601 numeric representation of the day of the week | 1 (for Monday) through 7 (for Sunday)
S | English ordinal suffix for the day of the month, 2 characters | st, nd, rd or th
w | Numeric representation of the day of the week | 0 (for Sunday) through 6 (for Saturday)
z | The day of the year (starting from 0) | 0 through 365


#### Week

character | Description | Example
:---------------:|-------------|--------
W | ISO-8601 week number of year, weeks starting on Monday | 42 (the 42nd week in the year)


#### Month

character | Description | Example
:---------------:|-------------|--------
F | A full textual representation of a month, such as January or March | January through December
m | Numeric representation of a month, with leading zeros | 01 through 12
M | A short textual representation of a month, three letters | Jan through Dec
n | Numeric representation of a month, without leading zeros | 1 through 12
t | Number of days in the given month | 28 through 31


#### Year

character | Description | Example
:---------------:|-------------|--------
L | Whether it's a leap year | 1 if it is a leap year, 0 otherwise.
o | ISO-8601 year number | 1999 or 2003
Y | A full numeric representation of a year, 4 digits | 1999 or 2003
y | A two digit representation of a year | Examples: 99 or 03


#### Time

character | Description | Example
:---------------:|-------------|--------
a | Lowercase Ante meridiem and Post meridiem | am or pm
A | Uppercase Ante meridiem and Post meridiem | AM or PM
B | Swatch Internet time | 000 through 999
g | 12-hour format of an hour without leading zeros | 1 through 12
G | 24-hour format of an hour without leading zeros | 0 through 23
h | 12-hour format of an hour with leading zeros | 01 through 12
H | 24-hour format of an hour with leading zeros | 00 through 23
i | Minutes with leading zeros | 00 to 59
s | Seconds, with leading zeros | 00 through 59
u | Microseconds | 654000


#### Timezone

character | Description | Example
:---------------:|-------------|--------
e | Timezone identifier | UTC, GMT
I<br>uppercase "i" | Whether or not the date is in daylight saving time | 1 if in DST, 0 otherwise.
O | Difference to Greenwich time (GMT) in hours | +0200
P | Difference to Greenwich time (GMT) with colon between hours and minutes | +02:00
T | Timezone abbreviation | EST, MDT ...
Z | Timezone offset in seconds | -43200 through 50400


#### Full Date/Time

character | Description | Example
:---------------:|-------------|--------
c | ISO 8601 date | 2004-02-12T15:19:21+00:00
r | RFC 2822 formatted date | Thu, 21 Dec 2000 16:01:07 +0200
U | Seconds since the Unix Epoch (January 1 1970 00:00:00 GMT) | 1420108212


CHANGELOG.md
# Changelog

## v1.0.2

* Updating dependencies
* Adding more NodeJS versions in unit tests

## v1.0.1

* Tests fails with America/Montreal
* Coverage is pushed to Code Climate

## v1.0.0

* First release


tests/format.js

var year = 2015;
var month = 1;
var day = 1;
var hour = 10;
var minute = 30;
var second = 12;

var date = new Date(Date.UTC(year, month - 1, day, hour, minute, second));

QUnit.test('Test method "format" / year', function (assert) {
  assert.equal(date.format('Y'), year, '`Y` / A full numeric representation of a year, 4 digits');
  assert.equal(date.format('y'), year % 100, '`y` / A two digit representation of a year');

  assert.strictEqual(date.format('L'), '0', '`L` / Whether it\'s a leap year	1 if it is a leap year, 0 otherwise');
  assert.strictEqual(new Date(1900, 0).format('L'), '0', '`L` / 1900 is not a leap year (1900 % 100 === 0)');
  assert.strictEqual(new Date(2008, 0).format('L'), '1', '`L` / 2008 is & leap year (2008 % 100 !== 0 && 2008 % 4 === 0)');
  assert.strictEqual(new Date(2000, 0).format('L'), '1', '`L` / 2000 is & leap year (2000 % 400 === 0)');

  assert.equal(date.format('o'), '2015', '`o` / ISO-8601 year number');
  assert.equal(new Date(2016, 0).format('o'), '2015', '`o` / 1st of 2016 is in the last week of 2015');
});

QUnit.test('Test method "format" / month', function (assert) {
  assert.equal(date.format('m'), '01', '`m` / Numeric representation of a month, with leading zeros');
  assert.equal(date.format('n'), '1', '`n` / Numeric representation of a month, without leading zeros');

  assert.equal(date.format('M'), 'Jan', '`M` / A short textual representation of a month, three letters');
  assert.equal(new Date(2015, 1).format('M'), 'Feb', '`M` / February');
  assert.equal(new Date(2015, 2).format('M'), 'Mar', '`M` / March');
  assert.equal(new Date(2015, 3).format('M'), 'Apr', '`M` / April');
  assert.equal(new Date(2015, 4).format('M'), 'May', '`M` / May');
  assert.equal(new Date(2015, 5).format('M'), 'Jun', '`M` / June');
  assert.equal(new Date(2015, 6).format('M'), 'Jul', '`M` / July');
  assert.equal(new Date(2015, 7).format('M'), 'Aug', '`M` / August');
  assert.equal(new Date(2015, 8).format('M'), 'Sep', '`M` / September');
  assert.equal(new Date(2015, 9).format('M'), 'Oct', '`M` / October');
  assert.equal(new Date(2015, 10).format('M'), 'Nov', '`M` / November');
  assert.equal(new Date(2015, 11).format('M'), 'Dec', '`M` / December');

  assert.equal(date.format('F'), 'January', '`F` / A full textual representation of a month, such as January or March');
  assert.equal(new Date(2015, 1).format('F'), 'February', '`F` / February');
  assert.equal(new Date(2015, 2).format('F'), 'March', '`F` / March');
  assert.equal(new Date(2015, 3).format('F'), 'April', '`F` / April');
  assert.equal(new Date(2015, 4).format('F'), 'May', '`F` / May');
  assert.equal(new Date(2015, 5).format('F'), 'June', '`F` / June');
  assert.equal(new Date(2015, 6).format('F'), 'July', '`F` / July');
  assert.equal(new Date(2015, 7).format('F'), 'August', '`F` / August');
  assert.equal(new Date(2015, 8).format('F'), 'September', '`F` / September');
  assert.equal(new Date(2015, 9).format('F'), 'October', '`F` / October');
  assert.equal(new Date(2015, 10).format('F'), 'November', '`F` / November');
  assert.equal(new Date(2015, 11).format('F'), 'December', '`F` / December');

  assert.equal(date.format('t'), 31, '`t` / Number of days in the given month');
  assert.equal(new Date(2015, 1).format('t'), 28, '`t` / February');
  assert.equal(new Date(2016, 1).format('t'), 29, '`t` / February (leap year)');
  assert.equal(new Date(2015, 2).format('t'), 31, '`t` / March');
  assert.equal(new Date(2015, 3).format('t'), 30, '`t` / April');
  assert.equal(new Date(2015, 4).format('t'), 31, '`t` / May');
  assert.equal(new Date(2015, 5).format('t'), 30, '`t` / June');
  assert.equal(new Date(2015, 6).format('t'), 31, '`t` / July');
  assert.equal(new Date(2015, 7).format('t'), 31, '`t` / August');
  assert.equal(new Date(2015, 8).format('t'), 30, '`t` / September');
  assert.equal(new Date(2015, 9).format('t'), 31, '`t` / October');
  assert.equal(new Date(2015, 10).format('t'), 30, '`t` / November');
  assert.equal(new Date(2015, 11).format('t'), 31, '`t` / December');
});

QUnit.test('Test method "format" / week', function (assert) {
  assert.equal(date.format('W'), 1, '`W` / Week of 1st day of 2015 => 1');
  assert.equal(new Date(2016, 0).format('W'), 53, '`W` / Week of 1st day of 2016 => 53');
});

QUnit.test('Test method "format" / day', function (assert) {
  var thursday = date;
  var friday = new Date(2015, 0, 2);
  var saturday = new Date(2015, 0, 3);
  var sunday = new Date(2015, 0, 4);
  var monday = new Date(2015, 0, 12);
  var tuesday = new Date(2015, 0, 13);
  var wednesday = new Date(2015, 0, 14);

  assert.equal(monday.format('d'), '12', '`d` / Day of the month, 2 digits with leading zeros / Monday');
  assert.equal(tuesday.format('d'), '13', '`d` / Day of the month, 2 digits with leading zeros / Tuesday');
  assert.equal(wednesday.format('d'), '14', '`d` / Day of the month, 2 digits with leading zeros / Wednesday');
  assert.equal(thursday.format('d'), '01', '`d` / Day of the month, 2 digits with leading zeros / Thursday');
  assert.equal(friday.format('d'), '02', '`d` / Day of the month, 2 digits with leading zeros / Friday');
  assert.equal(saturday.format('d'), '03', '`d` / Day of the month, 2 digits with leading zeros / Saturday');
  assert.equal(sunday.format('d'), '04', '`d` / Day of the month, 2 digits with leading zeros / Sunday');

  assert.equal(monday.format('D'), 'Mon', '`D` / A textual representation of a day, three letters / Monday');
  assert.equal(tuesday.format('D'), 'Tue', '`D` / A textual representation of a day, three letters / Tuesday');
  assert.equal(wednesday.format('D'), 'Wed', '`D` / A textual representation of a day, three letters / Wednesday');
  assert.equal(thursday.format('D'), 'Thu', '`D` / A textual representation of a day, three letters / Thursday');
  assert.equal(friday.format('D'), 'Fri', '`D` / A textual representation of a day, three letters / Friday');
  assert.equal(saturday.format('D'), 'Sat', '`D` / A textual representation of a day, three letters / Saturday');
  assert.equal(sunday.format('D'), 'Sun', '`D` / A textual representation of a day, three letters / Sunday');

  assert.equal(monday.format('j'), '12', '`j` / Day of the month without leading zeros / Monday');
  assert.equal(tuesday.format('j'), '13', '`j` / Day of the month without leading zeros / Tuesday');
  assert.equal(wednesday.format('j'), '14', '`j` / Day of the month without leading zeros / Wednesday');
  assert.equal(thursday.format('j'), '1', '`j` / Day of the month without leading zeros / Thursday');
  assert.equal(friday.format('j'), '2', '`j` / Day of the month without leading zeros / Friday');
  assert.equal(saturday.format('j'), '3', '`j` / Day of the month without leading zeros / Saturday');
  assert.equal(sunday.format('j'), '4', '`j` / Day of the month without leading zeros / Sunday');

  assert.equal(monday.format('l'), 'Monday', '`l` / A full textual representation of the day of the week / Monday');
  assert.equal(tuesday.format('l'), 'Tuesday', '`l` / A full textual representation of the day of the week / Tuesday');
  assert.equal(wednesday.format('l'), 'Wednesday', '`l` / A full textual representation of the day of the week / Wednesday');
  assert.equal(thursday.format('l'), 'Thursday', '`l` / A full textual representation of the day of the week / Thursday');
  assert.equal(friday.format('l'), 'Friday', '`l` / A full textual representation of the day of the week / Friday');
  assert.equal(saturday.format('l'), 'Saturday', '`l` / A full textual representation of the day of the week / Saturday');
  assert.equal(sunday.format('l'), 'Sunday', '`l` / A full textual representation of the day of the week / Sunday');

  assert.equal(monday.format('N'), '1', '`N` / ISO-8601 numeric representation of the day of the week / Monday / 1');
  assert.equal(tuesday.format('N'), '2', '`N` / ISO-8601 numeric representation of the day of the week / Tuesday / 2');
  assert.equal(wednesday.format('N'), '3', '`N` / ISO-8601 numeric representation of the day of the week / Wednesday / 3');
  assert.equal(thursday.format('N'), '4', '`N` / ISO-8601 numeric representation of the day of the week / Thursday / 4');
  assert.equal(friday.format('N'), '5', '`N` / ISO-8601 numeric representation of the day of the week / Friday / 5');
  assert.equal(saturday.format('N'), '6', '`N` / ISO-8601 numeric representation of the day of the week / Saturday / 6');
  assert.equal(sunday.format('N'), '7', '`N` / ISO-8601 numeric representation of the day of the week / Sunday / 7');

  assert.equal(monday.format('S'), 'th', '`S` / English ordinal suffix for the day of the month');
  assert.equal(tuesday.format('S'), 'th', '`S` / English ordinal suffix for the day of the month');
  assert.equal(wednesday.format('S'), 'th', '`S` / English ordinal suffix for the day of the month');
  assert.equal(thursday.format('S'), 'st', '`S` / English ordinal suffix for the day of the month');
  assert.equal(friday.format('S'), 'nd', '`S` / English ordinal suffix for the day of the month');
  assert.equal(saturday.format('S'), 'rd', '`S` / English ordinal suffix for the day of the month');
  assert.equal(sunday.format('S'), 'th', '`S` / English ordinal suffix for the day of the month');

  assert.equal(monday.format('w'), 1, '`w` / Numeric representation of the day of the week / Monday / 1');
  assert.equal(tuesday.format('w'), 2, '`w` / Numeric representation of the day of the week / Tuesday / 2');
  assert.equal(wednesday.format('w'), 3, '`w` / Numeric representation of the day of the week / Wednesday / 3');
  assert.equal(thursday.format('w'), 4, '`w` / Numeric representation of the day of the week / Thursday / 4');
  assert.equal(friday.format('w'), 5, '`w` / Numeric representation of the day of the week / Friday / 5');
  assert.equal(saturday.format('w'), 6, '`w` / Numeric representation of the day of the week / Saturday / 6');
  assert.equal(sunday.format('w'), 0, '`w` / Numeric representation of the day of the week / Sunday / 0');

  assert.equal(monday.format('z'), 11, '`z` / The day of the year (starting from 0) / Monday');
  assert.equal(tuesday.format('z'), 12, '`z` / The day of the year (starting from 0) / Tuesday');
  assert.equal(wednesday.format('z'), 13, '`z` / The day of the year (starting from 0) / Wednesday');
  assert.equal(thursday.format('z'), 0, '`z` / The day of the year (starting from 0) / Thursday');
  assert.equal(friday.format('z'), 1, '`z` / The day of the year (starting from 0) / Friday');
  assert.equal(saturday.format('z'), 2, '`z` / The day of the year (starting from 0) / Saturday');
  assert.equal(sunday.format('z'), 3, '`z` / The day of the year (starting from 0) / Sunday');

  assert.equal((new Date(2016, 0, 0)).format('z'), 364, '`z` / The day of the year (starting from 0) / Last day of 2015');
  assert.equal((new Date(2001, 0, 0)).format('z'), 365, '`z` / The day of the year (starting from 0) / Last day of 2000 (leap year)');
});

QUnit.test('Test method "format" / time', function (assert) {
  var amDate = new Date(2015, 0, 1, 9, 59, 3, 12);
  var pmDate = new Date(2015, 0, 1, 20, 9, 40, 345);

  assert.equal(amDate.format('a'), 'am', '`a` / Lowercase Ante meridiem and Post meridiem	/ am');
  assert.equal(pmDate.format('a'), 'pm', '`a` / Lowercase Ante meridiem and Post meridiem	/ pm');

  assert.equal(amDate.format('A'), 'AM', '`A` / Uppercase Ante meridiem and Post meridiem	/ AM');
  assert.equal(pmDate.format('A'), 'PM', '`A` / Uppercase Ante meridiem and Post meridiem	/ PM');

//  assert.equal(amDate.format('B'), 374, '`B` / Swatch Internet time / Morning');
//  assert.equal(pmDate.format('B'), 798, '`B` / Swatch Internet time / Afternoon');

  assert.equal(amDate.format('g'), 9, '`g` / 12-hour format of an hour without leading zeros / Morning');
  assert.equal(pmDate.format('g'), 8, '`g` / 12-hour format of an hour without leading zeros / Afternoon');

  assert.equal(amDate.format('G'), 9, '`G` / 24-hour format of an hour without leading zeros / Morning');
  assert.equal(pmDate.format('G'), 20, '`G` / 24-hour format of an hour without leading zeros / Afternoon');

  assert.equal(amDate.format('h'), '09', '`h` / 12-hour format of an hour with leading zeros / Morning');
  assert.equal(pmDate.format('h'), '08', '`h` / 12-hour format of an hour with leading zeros / Afternoon');

  assert.equal(amDate.format('H'), '09', '`H` / 24-hour format of an hour with leading zeros / Morning');
  assert.equal(pmDate.format('H'), '20', '`H` / 24-hour format of an hour with leading zeros / Afternoon');

  assert.equal(amDate.format('i'), '59', '`i` / Minutes with leading zeros / Morning');
  assert.equal(pmDate.format('i'), '09', '`i` / Minutes with leading zeros / Afternoon');

  assert.equal(amDate.format('s'), '03', '`s` / Seconds with leading zeros / Morning');
  assert.equal(pmDate.format('s'), '40', '`s` / Seconds with leading zeros / Afternoon');

  assert.equal(amDate.format('u'), '012000', '`s` / Microseconds / Morning');
  assert.equal(pmDate.format('u'), '345000', '`s` / Microseconds / Afternoon');
});
/*
QUnit.test('Test method "format" / Timezone', function (assert) {
  assert.equal(date.format('e'), 'CET', '`e` / Timezone identifier');

  assert.equal(date.format('I'), '0', '`I` / Is in DST / false');
  assert.equal((new Date(2015, 3)).format('I'), '1', '`I` / Is in DST / April 1st is true in France');

  assert.equal(date.format('O'), date.format('I') === '1' ? '+0200' : '+0100', '`O` / Difference to Greenwich time');

  assert.equal(date.format('P'), date.format('I') === '1' ? '+02:00' : '+01:00', '`P` / Difference to Greenwich time wtih colon');

  assert.equal(date.format('T'), 'CET', '`T` / Timezone abbreviation');

  assert.equal(date.format('Z'), date.format('I') === '1' ? 7200 : 3600, '`Z` / Timezone offset in seconds');
});
*/
QUnit.test('Test method "format" / Full date&time', function (assert) {
//  assert.equal(date.format('c'), '2015-01-01T11:30:12+01:00', '`c` / ISO 8601');
//  assert.equal(date.format('r'), 'Thu, 1 Jan 2015 11:30:12 +0100', '`r` / RFC 2822');
  assert.equal(date.format('U'), 1420108212, '`U` / Seconds since the Unix Epoch');
});

QUnit.test('Test method "format" / Other', function (assert) {
//  assert.equal(date.format('\\l\\e d/m/Y \\à H:i:s'), 'le 01/01/2015 à 11:30:12', 'Composed');
  assert.equal(date.format('\\\\d\\d'), '\\01d', 'Escaping the escaper');
  assert.equal(date.format('F jS'), 'January 1st', 'January 1st');
  assert.equal((new Date(2015, 4, 4)).format('F \\t\\h\\e jS \\b\\e \\y\\o\\u'), 'May the 4th be you', 'Star Wars day');
});


tests/isValid.js

QUnit.test('Test method "isValid"', function (assert) {
  assert.ok(new Date().isValid(), 'new Date() is valid');
  assert.ok(new Date('2015-01-01').isValid(), 'new Date(\'2015-01-01\') is valid');
  assert.ok(new Date(2015, 0, 1).isValid(), 'new Date(2015, 0, 1) is valid');
  assert.ok(!(new Date('lol').isValid()), 'new Date(\'lol\') is invalid');
});


Gruntfile.js
module.exports = function(grunt) {

  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),

    connect: {
      server: {
        options: {
          hostname: '127.0.0.1',
          base: '.',
          port: 8080
        }
      }
    },

    jshint: {
      files: ['src/*.js'],
      options: {
        jshintrc: '.jshintrc'
      }
    },

    uglify: {
      grouped: {
        options: {
          mangle: false,
          compress: false,
          beautify: true
        },
        files: {
          'build/date-enhancer.js': ['src/*.js']
        }
      },
      minified: {
        options: {
          mangle: true,
          compress: {
            sequences: true,
            properties: true,
            dead_code: true,
            booleans: true,
            loops: true,
            if_return: true,
            join_vars: true,
            drop_console: true,
          }
        },
        files: {
          'build/date-enhancer.min.js': ['src/*.js']
        }
      }
    },

    jasmine : {
      src : 'src/*.js',
      options : {
        specs : 'spec/*.js'
      }
    },

    watch: {
      scripts: {
        files: ['src/*', 'spec/*', 'Gruntfile.js'],
        tasks: ['default'],
        options: {
          interrupt: true,
        },
      },
    },

    karma: {
      options: {
        frameworks: ['jasmine'],
        files: ['spec/*.js', 'src/*.js', 'spec/env/' + (process.env.tz ? process.env.tz : 'unknown') + '.js'],
      },
      unit: {
        browsers: ['PhantomJS', 'Safari', 'Firefox'],
      },
      travis: {
        browsers: ['PhantomJS', 'Firefox'],
        singleRun: true,
        reporters: ['progress', 'coverage'],
        coverageReporter: {
          type: "lcov",
          dir: "coverage/"
        },
        preprocessors: {
          "src/*js": "coverage"
        },
      }
    },

    coveralls: {
      options: {
        coverageDir: 'coverage/',
        force: true,
        recursive: true
      }
    }
  });

  grunt.loadNpmTasks('grunt-contrib-connect');
  grunt.loadNpmTasks('grunt-contrib-jshint');
  grunt.loadNpmTasks('grunt-contrib-uglify');
  grunt.loadNpmTasks('grunt-contrib-watch');
  grunt.loadNpmTasks('grunt-contrib-jasmine');
  grunt.loadNpmTasks('grunt-karma');
  grunt.loadNpmTasks('grunt-karma-coveralls');

  grunt.registerTask('travis', ['jshint', 'karma:travis', 'coveralls']);
  grunt.registerTask('default', ['karma:unit', 'jshint', 'uglify']);
};


src/format.js
(function () {
  'use strict';

  if ('function' !== Date.prototype.format) {
    Date.locale = {};
    Date.locale.en = {
      months: {
        longs: [
          'January', 'February', 'March', 'April', 'May', 'June',
          'July', 'August', 'September', 'October', 'November', 'December'
        ],
        shorts: ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec']
      },
      days: {
        longs: ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'],
        shorts: ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat']
      },
      suffixes: ['th', 'st', 'nd', 'rd']
    };

    Date.locale.use = Date.locale.use || 'en';

    Date.prototype.format = function (pattern) {
      var leadingZeros = function (string, length) {
        length = typeof length === 'undefined' ? 2 : length;
        string = string.toString();

        while (string.length < length) {
          string = '0' + string;
        }

        return string;
      };

      var charValues = {

        // Year
        Y: function () { // Year, 4 digits
          return this.getFullYear();
        },
        y: function () { // Year, 2 digits
          return this.getFullYear() % 100;
        },
        L: function () { // Is a leap year
          var year = this.getFullYear();

          if ((year % 4 === 0 && year % 100 !== 0) || (year % 400 === 0)) {
            return 1;
          }

          return 0;
        },
        o: function () { // Week year
          return Math.floor(this.getWeekNumber() / 100);
        },

        // Month
        n: function () { // No leading zero
          return this.getMonth() + 1;
        },
        m: function () { // With leading zero
          return leadingZeros(this.format('n'));
        },
        M: function () { // 3 letters
          return Date.locale[ Date.locale.use ].months.shorts[ this.getMonth() ];
        },
        F: function () { // Full text
          return Date.locale[ Date.locale.use ].months.longs[ this.getMonth() ];
        },
        t: function () { // Nb of days
          return new Date(this.getFullYear(), this.getMonth() + 1, 0).getDate();
        },

        // Weeks
        W: function () { // Week number
          return this.getWeekNumber() % 100;
        },

        // Days
        d: function () { // With leading zero
          return leadingZeros(this.getDate());
        },
        j: function () { // Without leading zero
          return this.getDate();
        },
        D: function () { // 3 letters
          return Date.locale[ Date.locale.use ].days.shorts[ this.getDay() ];
        },
        l: function () { // Full text
          return Date.locale[ Date.locale.use ].days.longs[ this.getDay() ];
        },
        N: function () { // Day of week ISO
          var dow = this.getDay();

          return 0 === dow ? 7 : dow;
        },
        S: function () { // Ordinal suffix
          var suffix = Date.locale[ Date.locale.use ].suffixes[ this.getDate() ];

          return typeof suffix === 'undefined' ? Date.locale[ Date.locale.use ].suffixes[0] : suffix;
        },
        w: function () { // Day of week
          return this.getDay();
        },
        z: function () { // Day of year
          var ref = new Date(this.getFullYear(), 0, 0);
          var elapsed = this.getTime() - ref.getTime();

          return Math.floor(elapsed / 864e5) - 1;
        },

        // Time
        a: function () { // am / pm
          return this.getHours() < 12 ? 'am' : 'pm';
        },
        A: function () { // am / pm
          return this.format('a').toUpperCase();
        },
        B: function () { // swatch internet time
          var hours   = (this.getUTCHours() + 1) * 3600;
          var minutes = this.getUTCMinutes() * 60;
          var seconds = this.getUTCSeconds();

          return leadingZeros(Math.floor((seconds + minutes + hours) / 86.4) % 1000, 3);
        },
        g: function () { // Hours, 12h, without leading zeros
          return this.getHours() % 12;
        },
        G: function () { // Hours, 24h, without leading zeros
          return this.getHours();
        },
        h: function () { // Hours, 12h, with leading zeros
          return leadingZeros(this.format('g'));
        },
        H: function () { // Hours, 24h, with leading zeros
          return leadingZeros(this.getHours());
        },
        i: function () { // Minutes, with leading zeros
          return leadingZeros(this.getMinutes());
        },
        s: function () { // Seconds, with leading zeros
          return leadingZeros(this.getSeconds());
        },
        u: function () { // Microseconds
          return leadingZeros(this.getMilliseconds() * 1e3, 6);
        },

        // Timezone
        e: function () { // TZ Name
          return (new Date()).toString().match(/\((.+)\)/)[1];
        },
        I: function () { // Is in DST
          // Based on http://stackoverflow.com/a/11888430

          var january = new Date(this.getFullYear(), 0);
          var july = new Date(this.getFullYear(), 6);

          var maxOffset = Math.max(january.getTimezoneOffset(), july.getTimezoneOffset());

          return this.getTimezoneOffset() !== maxOffset ? '1' : '0';
        },
        O: function () { // Diff to GMT +0000 style
          return this.format('P').replace(':', '');
        },
        P: function () { // Diff to GMT +00:00 style
          var offset = this.getTimezoneOffset();
          var sign = offset <= 0 ? '+' : '-';

          offset = Math.abs(offset);

          return sign + leadingZeros(Math.floor(offset / 60)) + ':' + leadingZeros(offset % 60);
        },
        T: function () { // TZ Abbr, same as TZ name, JS is a shame on TZ manipulation
          return this.format('e');
        },
        Z: function () { // TZ offset in seconds
          return -60 * this.getTimezoneOffset();
        },

        // Full Date/Time
        c: function () { // ISO 8601
          return this.format('Y-m-d\\TH:i:sP');
        },
        r: function () { // RFC 2822
          return this.format('D, j M Y H:i:s O');
        },
        U: function () { // UNIX Timestamp
          return Math.floor(this.getTime() / 1000);
        }
      };
      var that = this;

      return pattern.replace(/\\?(.)/g, function (character) {
        return charValues[character] ? charValues[character].call(that) : character.replace(/^\\(.+)/, '$1');
      });
    };
  }
})();
